contracts/src/registry/EnhancedAccessControl.sol

```diff
@@ -0,0 +1,205 @@
+     * @param account The account to check.
+     * @return `true` if `account` has been granted all the given roles in either `resource` or the `ROOT_RESOURCE`, `false` otherwise.
+     */
+    function hasRoles(bytes32 resource, uint256 rolesBitmap, address account) public view virtual returns (bool) {
+        return ((_roles[resource][account] | _roles[ROOT_RESOURCE][account]) & rolesBitmap) == rolesBitmap;
+    }
+
+
+    /**
+     * @dev Grants roles to `account`.
+     *
+     * The caller must have all the necessary admin roles for the roles being granted.
+     *
+     * @param resource The resource to grant roles within.
+     * @param roleBitmap The roles bitmap to grant.
+     * @param account The account to grant roles to.
+     * @return `true` if the roles were granted, `false` otherwise.
+     */
+    function grantRoles(bytes32 resource, uint256 roleBitmap, address account) public virtual canGrantRoles(resource, roleBitmap) returns (bool) {
+        return _grantRoles(resource, roleBitmap, account);
+    }
+
+
+    /**
+     * @dev Revokes roles from `account`.
+     *
+     * The caller must have all the necessary admin roles for the roles being revoked.
+     *
+     * @param resource The resource to revoke roles within.
+     * @param roleBitmap The roles bitmap to revoke.
+     * @param account The account to revoke roles from.
+     * @return `true` if the roles were revoked, `false` otherwise.
+     */
+    function revokeRoles(bytes32 resource, uint256 roleBitmap, address account) public virtual canGrantRoles(resource, roleBitmap) returns (bool) {
+        return _revokeRoles(resource, roleBitmap, account);
+    }
+
+    // Internal functions
+
+    /**
+     * @dev Reverts if `account` does not have all the given roles.
+     */
+    function _checkRoles(bytes32 resource, uint256 roleBitmap, address account) internal view virtual {
+        if (!hasRoles(resource, roleBitmap, account)) {
+            revert EACUnauthorizedAccountRoles(resource, roleBitmap, account);
+        }
+    }
+
+    /**
+     * @dev Reverts if `account` does not have the admin roles for all the given roles.
+     */
+    function _checkCanGrantRoles(bytes32 resource, uint256 roleBitmap, address account) internal view virtual {
+        uint256 adminRoles = roleBitmap & 0xffffffffffffffffffffffffffffffff00000000000000000000000000000000;
+        if (adminRoles == 0 || !hasRoles(resource, adminRoles, account)) {
+            revert EACUnauthorizedAccountAdminRoles(resource, roleBitmap, account);
+        }
+    }
+
+    /**
+     * @dev Copies all roles from `srcAccount` to `dstAccount` within the same resource.
+     *
+     * @param resource The resource to copy roles within.
+     * @param srcAccount The account to copy roles from.
+     * @param dstAccount The account to copy roles to.
+     */
+    function _copyRoles(bytes32 resource, address srcAccount, address dstAccount) internal virtual {
```

Comment:
Perhaps there should be an easy way to *move* roles, too? We need to consider if we want to allow users to move roles to another account even if they aren't admin for those roles, as well.

Reply:
This is an internal method that can be called by subclasses  and surrounded by whatever rules they want. To do a move a subclass could simply copy/grant followed by a revoke. I don't think we need to add anything more for this.